### PR TITLE
Creation API

### LINKED ISSUE #3: Define API for upstream users
Define user API.

### PR DESCRIPTION
**Pull request by sohkai**
Fixes #3 and #4.

Depends on #8, but changes there will be rebased onto here as necessary.

Implements:

- COALA IP models
- COALA IP creation functionality
- Plugin discovery, ala `setuptool`'s entry points
- Tests for the above with a mocked plugin

TODO:

- [x] Handle error cases
- [ ] Add more `docs/` documentation

### CODE REVIEW

**Comment by sohkai** [comment]
@TimDaub @diminator @vrde The code's not quite ready for review, but just wanted a first pass over the API to make sure everything's sane:

- The package exposes a `coalaip.models` module that contains the entity models (which just marshall around some given data with light sanity checks and linked data sprinkling)
- The actual module people should interact with is `coalaip.coalaip` (exported through `__init__.py`), which has higher-level functions that take in data and return instances of the models. These include:
   - `create_user`: Create and return a new user on the given plugin
   - `register_manifestation`: Register a new manifestation, returning the automatically registered Copyright, Manifestation, and Creation models
   - `derive_right`: Derive a right from a Copyright (**can ignore for now**)
   - `transfer_right`: Transfer a right (derived or Copyright) (**can ignore for now**)
   - `bind_plugin`: Given a plugin, return a `coalaip` instance whose top level functions (the ones above) already have the plugin bound to them, for ease of use

**Comment by TimDaub** [comment]
In terms of architecture I'm really happy with how this looks. Great job :+1:

File: coalaip/coalaip.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            :class:`EntityCreationError`: if the manifestation, its
+                copyright, or the automatically created work (if no
+                existing work is given) fail to be created on the
+                persistence layer
+        """
+
+        # TODO: in the future, we may want to consider blocking (or asyncing) until
+        # we confirm that an entity has actually been created
+
+        # FIXME: is there a better way to do this? i.e. undefined in javascript
```
Comment (sohkai) [inline coalaip/coalaip.py]: Highlighting this for review

File: coalaip/models.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+                             'provided when instantiating a CoalaIp entity '
+                             'instance. Given a plugin of type '
+                             "'{}' instead.".format(type(plugin))))
+
+        if not isinstance(entity_type, str):
+            raise EntityDataError(('The entity type must be provided as a '
+                                   'string to the entity. '
+                                   'Got {} instead.').format(entity_type))
+
+        # FIXME: should I check that data is a dict?
```
Comment (sohkai) [inline coalaip/models.py]: Highlighting this for review.

Thinking about immediately throwing if `data` isn't a dict.

File: coalaip/models.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+        # If the entity type is already specified as part of the data, use that
+        # instead of 'CreativeWork'
+        for type_key in ['type', '@type']:
+            if type_key in data:
+                entity_type = data[type_key]
+                del data[type_key]
+                break
+
+        # FIXME: apply defaults
```
Comment (sohkai) [inline coalaip/models.py]: Nothing really comes to mind for defaults in Manifestations or Works, but maybe I missed something?

File: coalaip/models.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    """
+
+    def __init__(self, data, *, entity_type='Right', ctx=context_urls.COALAIP,
+                 plugin, **kwargs):
+        """Initialize a :class:`~coalaip.models.Right` instance
+
+        INSTANTIATION NOTES
+        """
+
+        # FIXME: apply defaults
```
Comment (sohkai) [inline coalaip/models.py]: Same for Rights, nothing right now comes to mind, but maybe there are some?

**Comment by sohkai** [comment]
Docs and further tests to be added later.

**Pull request merged**